lib/usd/translators/mayaReferenceUpdater.cpp

```diff
@@ -25,46 +25,86 @@
+     | UsdMayaPrimUpdater::Supports::AutoPull));
 PXRUSDMAYA_REGISTER_UPDATER(
     MayaUsd_SchemasALMayaReference,
+    reference,
     PxrUsdTranslators_MayaReferenceUpdater,
-    (UsdMayaPrimUpdater::Supports::Push | UsdMayaPrimUpdater::Supports::Clear));
+    (UsdMayaPrimUpdater::Supports::Push | UsdMayaPrimUpdater::Supports::Clear
+     | UsdMayaPrimUpdater::Supports::AutoPull));
 
 PxrUsdTranslators_MayaReferenceUpdater::PxrUsdTranslators_MayaReferenceUpdater(
     const MFnDependencyNode& depNodeFn,
-    const SdfPath&           usdPath)
-    : UsdMayaPrimUpdater(depNodeFn, usdPath)
+    const Ufe::Path&         path)
+    : UsdMayaPrimUpdater(depNodeFn, path)
 {
 }
 
 /* virtual */
-bool PxrUsdTranslators_MayaReferenceUpdater::Pull(UsdMayaPrimUpdaterContext* context)
+bool PxrUsdTranslators_MayaReferenceUpdater::PushCopySpecs(
+    SdfLayerRefPtr srcLayer,
+    const SdfPath& srcSdfPath,
+    SdfLayerRefPtr dstLayer,
+    const SdfPath& dstSdfPath)
 {
-    const UsdPrim& usdPrim = GetUsdPrim<MayaUsd_SchemasMayaReference>(*context);
-    const MObject& parentNode = GetMayaObject();
+    bool success = false;
+
+    // We are looking for a very specific configuration in here
+    // i.e. a parent prim with a variant set called "animVariant"
+    // and two variants "cache" and "rig"
+    UsdStageRefPtr srcStage = UsdStage::Open(dstLayer);
+    SdfPath        parentSdfPath = dstSdfPath.GetParentPath();
+    UsdPrim        primWithVariant = srcStage->GetPrimAtPath(parentSdfPath);
+
+    // Switching variant to cache to discover payload and where to write the data
+    UsdVariantSet variantSet = primWithVariant.GetVariantSet("animVariant");
+    variantSet.SetVariantSelection("cache");
 
-    UsdMayaTranslatorMayaReference::update(usdPrim, parentNode);
+    // Find the layer and prim to which we need to copy the content of push
+    // There is currently no easy way to query this information at prim level
+    // so we go to pcp and sdf.
+    UsdPrim                 primWithPayload = primWithVariant.GetChildren().front();
+    UsdPrimCompositionQuery query(primWithPayload);
+    for (const auto& arc : query.GetCompositionArcs()) {
+        if (arc.GetArcType() == PcpArcTypePayload) {
+            SdfLayerHandle payloadLayer
+                = arc.GetTargetNode().GetLayerStack()->GetIdentifier().rootLayer;
+            SdfPath payloadPrimPath = arc.GetTargetNode().GetPath();
+            success = SdfCopySpec(srcLayer, srcSdfPath, payloadLayer, payloadPrimPath);
 
-    return true;
+            break;
+        }
+    }
+
+    return success;
 }
 
 /* virtual */
-void PxrUsdTranslators_MayaReferenceUpdater::Clear(UsdMayaPrimUpdaterContext* context)
+bool PxrUsdTranslators_MayaReferenceUpdater::DiscardEdits(const UsdMayaPrimUpdaterContext& context)
 {
     const MObject& parentNode = GetMayaObject();
-
     UsdMayaTranslatorMayaReference::UnloadMayaReference(parentNode);
```

Comment:
This doesn't discard edits on MayaReference.

Reply:
Quite true.  If you feel this needs to be part of this pull request, I can have a look, otherwise we can consider implementing this in the next phase, which will be for Maya reference workflows.